### ISSUE TITLE
properly define what a "root" is

### ISSUE THREAD

**Issue by Gankra** [bholley/cargo-vet#70]
`cargo metadata` has a field for an explicit "root" package but this doesn't make sense for virtual workspaces, or anything with multiple artifacts and can be null.

I am currently using "workspace members" as an approximation of a root but this definitely includes things that *aren't* roots. I'm not sure if we want to definitely include those things "as" roots or actually only include things with "no reverse dependencies" as "roots".

Notably I think non-root members *can* have tests/benchmarks, so it depends on why exactly you *care* about the concept of a "root" package. Conversely it's non-obvious to me if some packages can be *not* in the workspace-members list and still have tests that get run. It's very unclear why cargo picks the particular packages it picks to be "members".

**Comment by bholley** [comment]
So per Matrix discussion, it sounds like cargo includes all first-party crates as workspace members, unless they're explicitly excluded (which m-c does haphazardly for various crates [1]). I guess the thinking here is that top-level `cargo test` should run tests on all first-party crates, which means they need to be buildable as root crates with their dev-dependencies.

It seems in general that a workspace will have (A) a few members that correspond to actual distributable products (either a binary like wasmtime, or a library that gets linked into a foreign binary like Firefox), and then (B) a bunch of other members that just correspond to the testing configuration for various first-party sub-crates.

I think 99% of the time you just want (B) to be safe-to-run: you're not shipping it, but you don't want a top-level invocation of `cargo test` to compromise the machine. And so ideally this would happen by default without `cargo vet` users needing to understand this intricacy of how workspaces work or writing a bunch of boilerplate policies.

I think "crates with no reverse-dependencies" is a good proxy for (A). It's not perfect — you might have one distributable first-party crate that depends on another (e.g. wasmtime and cranelift) — but I think it's probably a good enough starting point from which to customize policies.

So in summary, I think we can should define policy roots as "crates with no reverse-dependencies" and then relying on inherited `build-and-dev-criteria` (which defaults to `safe-to-run`) to achieve the right level of auditing for (B).

**Comment by bholley** [comment]
@Gankra anything left to do here? My sense is that the implementation is doing what I proposed so I think we can close this. Please reopen if not.
